types/types.go

```diff
@@ -26,63 +26,81 @@ import (
 // Marker helps to mark alerts as silenced and/or inhibited.
 // All methods are goroutine-safe.
 type Marker interface {
-	SetInhibited(alert model.Fingerprint, b bool)
 	SetSilenced(alert model.Fingerprint, sil ...string)
+	SetStatus(alert model.Fingerprint, status uint8, b bool)
 
 	Silenced(alert model.Fingerprint) (string, bool)
-	Inhibited(alert model.Fingerprint) bool
+	Status(alert model.Fingerprint, status uint8) bool
 }
 
 // NewMarker returns an instance of a Marker implementation.
 func NewMarker() Marker {
 	return &memMarker{
-		inhibited: map[model.Fingerprint]struct{}{},
-		silenced:  map[model.Fingerprint]string{},
+		silenced: map[model.Fingerprint]string{},
+		status:   map[model.Fingerprint]uint8{},
 	}
 }
 
+const (
+	// InhibitedStatus bit is set on memMarker->status when linked alert is
+	// inhibited
+	InhibitedStatus = 1
+	// ProcessedStatus bit is set on memMarker->status when linked alert is
+	// processed by notify code
+	ProcessedStatus = 2
+)
+
 type memMarker struct {
 	inhibited map[model.Fingerprint]struct{}
 	silenced  map[model.Fingerprint]string
+	status    map[model.Fingerprint]uint8
 
 	mtx sync.RWMutex
 }
 
-func (m *memMarker) Inhibited(alert model.Fingerprint) bool {
+func (m *memMarker) Silenced(alert model.Fingerprint) (string, bool) {
 	m.mtx.RLock()
 	defer m.mtx.RUnlock()
 
-	_, ok := m.inhibited[alert]
-	return ok
+	sid, ok := m.silenced[alert]
+	return sid, ok
 }
 
-func (m *memMarker) Silenced(alert model.Fingerprint) (string, bool) {
+func (m *memMarker) Status(alert model.Fingerprint, status uint8) bool {
 	m.mtx.RLock()
 	defer m.mtx.RUnlock()
 
-	sid, ok := m.silenced[alert]
-	return sid, ok
+	s, ok := m.status[alert]
+	if !ok {
+		return false
+	}
+	return s&status != 0
 }
 
-func (m *memMarker) SetInhibited(alert model.Fingerprint, b bool) {
+func (m *memMarker) SetSilenced(alert model.Fingerprint, sil ...string) {
 	m.mtx.Lock()
 	defer m.mtx.Unlock()
 
-	if !b {
-		delete(m.inhibited, alert)
+	if len(sil) == 0 {
+		delete(m.silenced, alert)
 	} else {
-		m.inhibited[alert] = struct{}{}
+		m.silenced[alert] = sil[0]
 	}
 }
 
-func (m *memMarker) SetSilenced(alert model.Fingerprint, sil ...string) {
+func (m *memMarker) SetStatus(alert model.Fingerprint, status uint8, b bool) {
 	m.mtx.Lock()
 	defer m.mtx.Unlock()
 
-	if len(sil) == 0 {
-		delete(m.silenced, alert)
+	if b {
+		m.status[alert] = m.status[alert] | status
 	} else {
-		m.silenced[alert] = sil[0]
+		m.status[alert] = m.status[alert] &^ status
+	}
+
+	// if there are no flags set remove the whole key
+	if m.status[alert] == 0 {
+		delete(m.status, alert)
```

Comment:
Looking for the code that removes Alerts ... that should also (hopefully) have access to the marker map, so we can get rid of the alert.